tests/testing-resources/include/aws/testing/AwsTestHelpers.h

```diff
@@ -197,13 +236,41 @@ class RetryPlanner
     private:
         using FunctionSuite = std::list<FunctionBlock>;
         FunctionSuite _suite;   
-        std::shared_ptr<CONTEXT> _contextSp;
+        CONTEXT_TYPE _contextSp;
         Aws::UnorderedSet<Aws::String> _uniqueFunctionBlockSet;
         size_t _cursor;        
 
         static const char ALLOCATION_TAG[];
 
+        //compile time conditional invoking of callable function with different arguments depending upon available context or not in c++ 11
+        
+        using RetryFunctionWrapped = std::function<bool (const RetryFunction_t& callable, const BlockIdentifierType& , size_t) >;
```

Comment:
Added a way for the Retry Planner class to not mandate context where test case doesn't need one. This is implemented using SFINAE concepts and type traits